voxel_grid/prototype/mvp_v1.hpp

```diff
@@ -0,0 +1,226 @@
+
+template <typename T>
+using get_grid_type = typename T::Grid;
+
+template <typename GridStruct, typename Grid = get_grid_type<GridStruct>, typename PointT = get_point_type<GridStruct>>
+class GridFilter : public Filter<PointT>, public GridStruct
+{
+    // TODO:
+    // check if the required functions exist in Grid
+    // static_assert(...);
+
+    using PointCloud = PointCloudT<PointT>;
+
+public:
+    GridFilter() : GridStruct()
+    {
+    }
+
+    const PointCloud &getInputCloud() const
+    {
+        return input_;
+    }
+
+    void setInputCloud(const PointCloud &input)
+    {
+        input_ = input;
+    }
+
+    Eigen::Vector3f getLeafSize() const
+    {
+        return leaf_size_;
+    }
+
+    void getMinMax3D(const PointCloud &cloud, Vec3f &min_pt, Vec3f &max_pt) const
+    {
+        // TODO: set initialize values
+        for (const auto &pt : cloud.points)
+        {
+            min_pt = min_pt.cwiseMin(pt.p);
+            max_pt = max_pt.cwiseMax(pt.p);
+        }
+    }
+
+protected:
+    void applyFilter(PointCloud &output)
+    {
+        GridStruct::setUp(this);
+        GridStruct::addPointsToGrid(input_, grid_);
+
+        for (auto it = grid_.begin(); it != grid_.end(); ++it)
+            if (std::optional<PointT> res = GridStruct::filterGrid(it, grid_); res)
+                output.push_back(res.value());
+    }
+
+    PointCloud input_;
+    Grid grid_;
+    Eigen::Vector3f leaf_size_ = Eigen::Vector3f::Constant(0.05);
+    // ...
+};
+
+// ================== VoxelGrid.hpp ==================
+
+// structure of a single voxel
+struct Voxel
+{
+    Vec3f centroid = Vec3f::Zero();
+    size_t num_pt = 0;
+};
+
+// structure and operations of the grid
+template <typename PointT>
+class VoxelStructT
+{
+public:
+    // Require user to define them
+    // not sure if there are a nicer way to do it
+    using PointCloud = PointCloudT<PointT>;
+    using Grid = std::unordered_map<size_t, Voxel>;
+
+    void setMinimumPointsNumberPerVoxel(const size_t num)
+    {
+        min_points_per_voxel_ = num;
+    }
+
+protected:
+    void setUp(const GridFilter<VoxelStructT> *grid_filter)
+    {
+        Vec3f min_p, max_p;
+        const auto &input = grid_filter->getInputCloud();
+        grid_filter->getMinMax3D(input, min_p, max_p);
+
+        inverse_leaf_size_ = 1 / grid_filter->getLeafSize().array();
+        min_b_ = (min_p.array() * inverse_leaf_size_.array()).cast<int>();
+        max_b_ = (max_p.array() * inverse_leaf_size_.array()).cast<int>();
+
+        div_b_ = (max_b_ - min_b_).array() + 1;
+        divb_mul_ << div_b_[0], div_b_[0], div_b_[1];
+    }
+
+    // only for accessing public methods
+    const auto getDerived()
+    {
+        return static_cast<GridFilter<VoxelStructT> *>(this);
+    }
+
+    size_t hashPoint(const PointT &pt)
+    {
+        const Vec3f tmp = pt.p.array() * inverse_leaf_size_.array();
+        return divb_mul_.dot(tmp.cast<int>() - min_b_);
```

Comment:
Would it not be enough to simply use uint32 or even better std::size_t consistently instead of int (int32) to prevent overflowing and support larger volumes/finer grid sizes?

Reply:
I am thinking to eliminate `getMinMax3D` (an `O(N)` operation) by using `tuple` and `unordered_map` together. In such case, we don't need `min_b_` and `max_b_` to [check overflow](https://github.com/PointCloudLibrary/pcl/blob/8ce5d698d23dc94476e8e080908ae80ccf6a2b0d/filters/include/pcl/filters/impl/voxel_grid.hpp#L241) and [compute hash](https://github.com/PointCloudLibrary/pcl/blob/8ce5d698d23dc94476e8e080908ae80ccf6a2b0d/filters/include/pcl/filters/impl/voxel_grid.hpp#L304).

It's just a small suggestion, I'm planing to use `size_t` in the new implementation as I don't want to bring too many code logic changes in one shot. This will probably enough for 99.9% of applications